### PR TITLE
✨ [Feature] Add Min/Max Gridding and Terrain Derivatives

### PR DESCRIPTION
**Pull request by lhemerly**
✨ **What**
This PR introduces two highly demanded features for point cloud and terrain processing:
1. **Min/Max Gridding (`algos/min_max.py`)**: Computes Digital Terrain Models (DTMs) retaining only the minimum or maximum Z values per cell. This is particularly useful for extracting bare-earth models (min) or canopy models (max) from raw point clouds. It leverages high-performance Taichi atomic operations (`ti.atomic_min`, `ti.atomic_max`).
2. **Terrain Derivatives (`algos/terrain_derivatives.py`)**: Adds the ability to rapidly calculate standard raster derivatives—Slope, Aspect, and Hillshade—directly from a numpy DTM array. This uses a parallel Taichi 3x3 window kernel optimized for large grids.

Both features have been exposed via `parapoint/__init__.py`.

✨ **Why**
These algorithms are standard tools in GIS and LiDAR processing workflows. Implementing them directly in Taichi within Parapoint allows users to keep their entire terrain processing pipeline highly parallelized and computationally efficient. 

✨ **Verification**
- Added `tests/test_min_max.py` to verify atomic minimum and maximum value selection from arrays.
- Added `tests/test_terrain_derivatives.py` to assert mathematical correctness (e.g. slope calculations) on standard edge-case dummy slopes.
- All 41 tests pass. Code formatted with Ruff.

---
*PR created automatically by Jules for task [16084403681084152525](https://jules.google.com/task/16084403681084152525) started by @lhemerly*

### CODE REVIEW

File: algos/min_max.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    count_np = count_field.to_numpy()
+
+    if return_min:
+        res_np = min_z_field.to_numpy()
+    else:
+        res_np = max_z_field.to_numpy()
+
+    dtm_np = np.full((grid_height, grid_width), nodata_value, dtype=np.float32)
+    valid_cells_mask = count_np > 0
+    dtm_np[valid_cells_mask.T] = res_np[valid_cells_mask]
```
[the remark this answers is not in the document]
Comment (lhemerly) [reply algos/min_max.py]: > **P1 Badge Preserve x/y mapping when writing min/max raster**
> 
> The masked assignment `dtm_np[valid_cells_mask.T] = res_np[valid_cells_mask]` uses different linearization orders on each side: the destination mask is transposed to `(y, x)` order, but the source values are still flattened in `(x, y)` order. As soon as valid cells span more than one row and column, min/max values can be written into the wrong raster positions, producing spatially misregistered outputs. Please write values with consistent indexing (for example by transposing `res_np` before masking or by explicit `(giy, gix)` indices).
> 
> Useful? React with 👍 / 👎.

@jules fix
